drivers/usb/device/usb_dc_dw.c

```diff
@@ -405,6 +405,8 @@ static int usb_dw_tx(u8_t ep, const u8_t *const data,
 		avail_space = usb_dw_tx_fifo_avail(ep_idx);
 		if (avail_space == usb_dw_ctrl.in_ep_ctrl[ep_idx].fifo_size) {
 			break;
+		} else {
+			usb_dw_flush_tx_fifo(ep_idx);
```

Comment:
Can you please explain why flush the fifo necessary here? What happens to the previous (maybe unprocessed) data?

Reply:
Can you please explain why flush the fifo necessary here?
>>  Because avilable fifo space is not same as the fifo size saved during IN endpoint configuration. If flushing is not done , then code will be in while(1) and hangs. This change was there previously but removed. You can refer to following commitid f591e582117c463ba8a3467110c7c9048d3967be

What happens to the previous (maybe unprocessed) data?
>> if complete data transfer not happened from device side then device don't respond. Because all the callbacks registered for IN endpoint is called only when transfer completed status bit is set.

Reply:
>Because avilable fifo space is not same as the fifo size saved during IN endpoint configuration. If flushing is not done , then code will be in while(1) and hangs

Why is it for ever not same as configured size? Can it be that:
- e.g. there is an error in the communication with the host, the fifo is not empty and endpoint is stalled and the host does not transmits IN tokens? But then the host will try to recover and clear stall. So it would be nice to have usb_dw_flush_tx_fifo(ep_idx) inside usb_dc_ep_clear_stall.
- there is a packet in the fifo and the controller waits for the IN token but somewhere a error has happened undiscovered? Whatever happened, just to empty fifo is not good, here fifo flashing should happen only after a timeout with error message.

>This change was there previously but removed. You can refer to following commitid f591e58

Yes it was removed by eb83cda7705e2e8402fc28bffb2871d3642e9dbf.